File: python/src/utils/generators.py
"""
Generator utilities for game entities
Demonstrates: Generators, async generators, type hints for iterators
"""

import asyncio
from typing import Iterator, AsyncIterator
from entities.obstacle import Obstacle


def obstacle_generator() -> Iterator[Obstacle]:
    """
    Infinite generator that produces random obstacles
    Demonstrates: Generator function with yield, Iterator type hint

    Yields:
        Obstacle: A new random obstacle

    Example:
        gen = obstacle_generator()
        obstacle1 = next(gen)
        obstacle2 = next(gen)
    """
    while True:
        yield Obstacle.create_random(0.0)


def obstacle_generator_n(count: int) -> Iterator[Obstacle]:
    """
    Generator that produces N obstacles
    Demonstrates: Finite generator with parameter

    Args:
        count: Number of obstacles to generate

    Yields:
        Obstacle: A new random obstacle
    """
    for _ in range(count):
        yield Obstacle.create_random(0.0)


async def async_obstacle_generator(delay_ms: int = 100) -> AsyncIterator[Obstacle]:
    """
    Async generator that produces obstacles with delay
    Demonstrates: Async generators (async def with yield), AsyncIterator type hint

    Args:
        delay_ms: Delay between obstacles in milliseconds

    Yields:
        Obstacle: A new random obstacle

    Example:
        async for obstacle in async_obstacle_generator(100):
            # Process obstacle
            pass
    """
    while True:
        await asyncio.sleep(delay_ms / 1000)
        yield Obstacle.create_random(0.0)


def take(iterable: Iterator, n: int) -> Iterator:
    """
    Iterator helper: Take first N items from an iterable
    Demonstrates: Generator composition, itertools-style patterns

    Args:
        iterable: Source iterable
        n: Number of items to take

    Yields:
        Items from the iterable

    Example:
        obstacles = take(obstacle_generator(), 5)
        # Gets exactly 5 obstacles
    """
    count = 0
    for item in iterable:
        if count >= n:
            break
        yield item
        count += 1

File: javascript/src/utils/generators.js
/**
 * Generator utilities for game entities
 * Demonstrates: Generator functions, infinite iterators, yield
 */

import { Obstacle } from '../entities/obstacle.js';

/**
 * Infinite generator that produces random obstacles
 * Demonstrates: Generator function (function*), yield
 *
 * @generator
 * @yields {Obstacle} - A new random obstacle
 *
 * @example
 * const gen = obstacleGenerator();
 * const obstacle1 = gen.next().value;
 * const obstacle2 = gen.next().value;
 */
export function* obstacleGenerator() {
  while (true) {
    yield Obstacle.createRandom(0);
  }
}

/**
 * Generator that produces N obstacles
 * Demonstrates: Finite generator with parameter
 *
 * @generator
 * @param {number} count - Number of obstacles to generate
 * @yields {Obstacle} - A new random obstacle
 */
export function* obstacleGeneratorN(count) {
  for (let i = 0; i < count; i++) {
    yield Obstacle.createRandom(0);
  }
}

/**
 * Async generator that produces obstacles with delay
 * Demonstrates: Async generators (async function*), for await...of
 *
 * @async
 * @generator
 * @param {number} delayMs - Delay between obstacles in milliseconds
 * @yields {Obstacle} - A new random obstacle
 */
export async function* asyncObstacleGenerator(delayMs = 100) {
  while (true) {
    await new Promise(resolve => setTimeout(resolve, delayMs));
    yield Obstacle.createRandom(0);
  }
}

/**
 * Iterator helper: Take first N items from an iterable
 * Demonstrates: Iterator pattern, generator composition
 *
 * @generator
 * @param {Iterable} iterable - Source iterable
 * @param {number} n - Number of items to take
 * @yields {*} - Items from the iterable
 */
export function* take(iterable, n) {
  let count = 0;
  for (const item of iterable) {
    if (count >= n) break;
    yield item;
    count++;
  }
}
